src/chapter_2/main.mo

```diff
@@ -76,4 +78,17 @@ actor chapter2{
           };
         };
       };
+
+      /// preupgrade function - converted the members hashmap to an array
+      system func preupgrade() {
+        stableMembers := Iter.toArray(members.entries());
+      };
+
+      /// postupgrade function to restore the map state by converting the array back to hashmap
+      system func postupgrade() {
+        members := HashMap.fromIter<Principal, Member>(
+        stableMembers.vals(), 1, Principal.equal, Principal.hash
+      );
+        stableMembers := [];
```

Comment:
any specific reason for setting stableMembers to an empty array?

Reply:
During the postupgrade, It restores the data from the stableMembers array back to the members HashMap.
Once the data is restored, stableMembers is no longer needed to hold this data because it has been transferred back to the main storage (members).